### PR TITLE
Implemented STP Tactics and the MoveTactic

### LINKED ISSUE #24: Create structure for STP logic
We need a basic implementation of our HL interface, which can be done with our old STP (Skills, Tactics, Plays) system. This module should return the Intents robots should run given the world state.

This is a large ticket that can have its work subdivided, but is left as a single ticket since it should be easier to view all these subtasks together, as they are related.

Module we need:
* PlayExecutor: Chooses what STP play to run and call that play
* Plays: Define when they are valid/applicable, and what tactics should be assigned with what priority to the available robots.
* Tactics: Define a single-robot role within the team
* Actions: Define a simple Action the robot can do, such a shoot or move

We should investigate and decide if we still want to use coroutines. If we choose to do so it will likely be asynchronous coroutines from boost for now.

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Implemented the Tactic structure for STP, and the MoveTactic.

### Testing Done

Unit tests for MoveTactic

### Resolved Issues

partially resolves #24 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/ai/hl/stp/tactic/move_tactic.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+#include "ai/hl/stp/action/move_action.h"
+#include "ai/hl/stp/tactic/tactic.h"
+
+class MoveTactic : public Tactic
+{
+   public:
+    explicit MoveTactic(const Robot& robot);
+
+    /**
+     * Returns the next Intent this MoveAction wants to run, given the parameters.
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/tactic/move_tactic.h]:
``` suggestion
     * Returns the next Intent this MoveTactic wants to run, given the parameters.
```

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+Tactic::Tactic(const Robot &robot)
+    : robot(robot),
+      intent_sequence(boost::bind(&Tactic::calculateNextIntentWrapper, this, _1))
+{
+}
+
+bool Tactic::done() const
+{
+    return !static_cast<bool>(intent_sequence);
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp:14]: comment plz

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+std::unique_ptr<Intent> Tactic::getNextIntent()
+{
+    // If the coroutine "iterator" is done, the calculateNextIntent function has completed
+    // and therefore the Tactic is done, so we return a null pointer
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp]:
``` suggestion
  // and therefore the Tactic is done, so we return an empty optional
```

Not super sure if `empty` is the right word here, but it's not really a null pointer

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.h [hunk 4]
```diff
@@ -17,24 +17,85 @@ class Tactic
...
+     * @return true if the Tactic is done and false otherwise
      */
-    virtual Robot selectRobot(const World& world,
-                              const std::vector<Robot>& available_robots) = 0;
+    bool done() const;
+
+    /**
+     * Evaluates the given Robot and returns a score indicating how optimal it would be
+     * for that Robot to perform this Tactic. A lower score indicates a more optimal
+     * robot, and returned score values must be >= 0.
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/tactic/tactic.h]: Can we make this a value in `[0,1]` so we have a concrete range?

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.h [hunk 5]
```diff
@@ -17,24 +17,85 @@ class Tactic
...
+     *
+     * @param robot The Robot to evaluate for this Tactic
+     *
+     * @return A score value >= 0 that indicates how optimal it would be for the given
+     * robot to perform this Tactic. Lower scores indicate more optimal/preferred robots.
+     */
+    virtual double evaluateRobot(const Robot& robot) = 0;
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/tactic/tactic.h]: maybe `calculateRobotQuality`?

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp (hunk 2 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp:14]: done

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.h (hunk 4 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/tactic/tactic.h]: Done

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.h (hunk 5 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/tactic/tactic.h]: name updated to `calculateRobotCost`

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp (hunk 2 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp:14]: nope

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.h (hunk 4 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/hl/stp/tactic/tactic.h]: nope

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.h (hunk 5 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/hl/stp/tactic/tactic.h]: nope

Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/tactic/tactic.h]: yup

File: src/thunderbots/software/ai/hl/stp/tactic/tactic.h (hunk 4 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/tactic/tactic.h]: yup

**Review by garethellis0 — APPROVED** [review]

**Pull request merged**